File: LeetCodeJava/src/practices/AddTwoNumbers.java
package practices;

public class AddTwoNumbers {
    public static ListNode add(ListNode l1, ListNode l2) {
        int carrier = 0;
        ListNode dump = new ListNode(0);
        ListNode curr = dump;

        while (l1!=null || l2!=null || carrier!=0) {
            int x = (l1!=null) ? l1.val:0;
            int y = (l2!=null) ? l2.val:0;
            int sum = carrier + x + y;
            carrier = sum / 10;
            curr.next = new ListNode(sum % 10);
            curr = curr.next;

            if (l1!=null) {
                l1 = l1.next;
            }

            if (l2!=null) {
                l2 = l2.next;
            }
        }

        return dump.next;
    }

    private static ListNode createLinkedList(int... values) {
        if (values==null || values.length==0) {
            return null;
        }
        ListNode head = new ListNode(values[0]);
        ListNode curr = head;
        for (int v : values) {
            curr.next = new ListNode(v);
            curr = curr.next;
        }
        return head;
    }

    private static void printLinkedList(ListNode node) {
        int i = 0;
        while (node!=null) {
            System.out.println("index: " + i + ", val: " + node.val);
            i++;
            node = node.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        ListNode l1 = createLinkedList(2, 4, 3);
        ListNode l2 = createLinkedList(5, 6, 4);
        ListNode ls12 = add(l1, l2);
        printLinkedList(ls12);

        ListNode l3 = createLinkedList(0);
        ListNode l4 = createLinkedList(0);
        ListNode l34 = add(l3, l4);
        printLinkedList(l34);

        ListNode l5 = createLinkedList(0, 9, 9, 9, 9, 9, 9, 9);
        ListNode l6 = createLinkedList(0, 9, 9, 9, 9);
        ListNode ls56 = add(l5, l6);
        printLinkedList(ls56);
    }
}

class ListNode {
    int val;
    ListNode next;

    ListNode() {

    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

File: LeetCodePython/src/practices/add_two_numbers.py
class ListNode:
    def __init__(self, val=0, nexxt=None):
        self.val = val
        self.nexxt = nexxt


def add_two_numbers(l1: ListNode, l2: ListNode) -> ListNode:
    carrier = 0
    dump: ListNode = ListNode(0)
    curr: ListNode = dump

    while l1 is not None or l2 is not None or carrier != 0:
        x = l1.val if l1 is not None else 0
        y = l2.val if l2 is not None else 0
        total = carrier + x + y
        carrier = total // 10
        curr.nexxt = ListNode(total % 10)
        curr = curr.nexxt
        l1 = l1.nexxt if l1 is not None else None
        l2 = l2.nexxt if l2 is not None else None

    return dump.nexxt


def _create_linked_list(values):
    if not values:
        return None
    head = ListNode(values[0])
    curr = head
    for val in values[1:]:
        curr.nexxt = ListNode(val)
        curr = curr.nexxt
    return head


def _print_linked_list(node):
    i = 0
    while node is not None:
        print(f"index: {i}, val: {node.val}")
        i += 1
        node = node.nexxt
    print()


if __name__ == '__main__':
    l1 = _create_linked_list([2, 4, 3])
    l2 = _create_linked_list([5, 6, 4])
    ls12 = add_two_numbers(l1, l2)
    _print_linked_list(ls12)

    l3 = _create_linked_list([0])
    l4 = _create_linked_list([0])
    ls34 = add_two_numbers(l3, l4)
    _print_linked_list(ls34)

    l5 = _create_linked_list([0, 9, 9, 9, 9, 9, 9, 9])
    l6 = _create_linked_list([0, 9, 9, 9, 9])
    ls56 = add_two_numbers(l5, l6)
    _print_linked_list(ls56)
